src/main/webapp/index-script.js

```diff
@@ -383,24 +383,114 @@ function degreesToRadians(degrees) {
+      // Timeout required to allow sortPlacesByDistance to finish.
+      setTimeout(function run() {resolve(places)}, 2000);
+    });
+    placePromise.then((places) => {
+      console.log("In the .then....");
+      console.log(places);
+      closeModal(document.getElementById("results-popup"));
+      triggerModal(document.getElementById("results-popup"));
+      populateResults(places);
+    });
+  });
+  
+  // Adds an event listener for the rating sort button
+  resultSortRating.addEventListener("click", () => { 
+    console.log("Clicked!!");
+    console.log(places);
+    const placePromise = new Promise((resolve, reject) => {
+      places = sortPlacesByRating(places, location);
+      console.log(places);
+      console.log("In the promise...");
+      // Timeout required to allow sortPlacesByRating to finish. 
+      setTimeout(function run() {resolve(places)}, 2000);
+    }).then((places) => {
+      console.log("In the .then....");
+      console.log(places);
+      closeModal(document.getElementById("results-popup"));
+      triggerModal(document.getElementById("results-popup"));
+      populateResults(places);
+    });
+  });
 }
 
-/* Adds a field 'rating' to each place with a random integer to
- * simulate sorting locations by their ratings. */
+/** Sorts place options by rating */
 function sortPlacesByRating(places) {
-  places.forEach((place) => {
-    let rand = Math.floor(Math.random() * 10);
-    place.rating = rand;
+  places.sort(function(a, b){
+    (getPlaceRating(a.place_id) > getPlaceRating(b.place_id)) ? 1 : -1});
+  return places;
+}
+
+/** Sorts place options by distance */
+function sortPlacesByDistance(places, current) {
+  const locationPromise = new Promise((resolve, reject) => {
+    // Check if they gave us a location to prioritize places near
+    if (current == null) {
+      // If not, check if we can rate by user's current location
+      if (navigator.geolocation) {
+        navigator.geolocation.getCurrentPosition(
+          position => {
+            const pos = {
+              lat: position.coords.latitude,
+              lng: position.coords.longitude
+            };
+          current = new google.maps.LatLng(pos.lat, pos.lng)
+          resolve(current);
+          }
+        );
+      } else {
+        // No location given, no sorting possible.
+        reject();
+      }
+    } else {
+      resolve(current);
+    }
+  }).then((locationResult) => {
+    console.log("Entered distance promise.then()...");
+    places.sort(function(a, b){ 
+      (getDistance(locationResult, a.geometry.location)
+        > getDistance(locationResult, b.geometry.location)) ? 1 : -1});
+    console.log("Sorted..");
+    return places;
   });
-  places.sort((a, b) =>
-    (a.rating > b.rating) ? 1 : -1);
+  // If I don't return here, I get type errors later on (line 511: can't read foreach of undefined)
   return places;
 }
 
+/** Returns the overall rating of a place */
+function getPlaceRating(placeID){
```

Comment:
why do we fetch the review again if we already used place.place_id to get the place ID here? Shouldn't the original instance already have place.rating?

Reply:
No, the original place that is called is directly from the Places API and does not contain any of the information that we have stored!

Reply:
To add on to that, this is the first time fetching the review (we don't fetch again unless one is clicked).

Reply:
Albeit a somewhat inefficient way to grab the rating, I think this is the best solution for getting the rating given our timeframe and current implementation.
My worry is this seems like a very costly process, do the ratings populate quickly?

Reply:
Yeah I couldn't think of a more efficient way of doing this - there isn't really a way to get our own personal rating without going to the backend and dragging it out. Since we don't have many reviews in here, this happens fairly quickly.

Reply:
Fair enough